### PR TITLE
Support pause (platform and lending pool level) and Settings

### PR DESCRIPTION
**Pull request by salazarguille**
It includes:

- A Settings contract to allow us:
     - To configure any required value in the platform.
     - To pause the platform.
     - To pause a lending pool.
- A Base contract which extends the ReentrancyGuard contract, and has the Settings contract.
- Changes in migration script.
- New unit tests to increase code coverage (almost 100%).

So,

Any contract can extend Base and get access to:
- Any setting defined in the Settings contract.
- Checking whether the platform or a lending pool is paused or not.

image

image

### CODE REVIEW

File: contracts/base/Base.sol [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        _;
+    }
+
+    /* Constructor */
+
+    /** External Functions */
+
+    /** Internal functions */
+
+    function initialize(address settingsAddress) internal isNotInitialized() {
```
Comment (hensha256) [inline contracts/base/Base.sol:64]: `_` missing from function name

File: contracts/base/Initializable.sol [hunk 2]
```diff
@@ -43,6 +43,10 @@ contract Initializable {
 
     /** External Functions */
 
+    function initialized() external view returns (bool) {
```
Comment (hensha256) [inline contracts/base/Initializable.sol:46]: Is there a reason to make this function? And not just make the variable public?

File: contracts/base/InterestConsensus.sol [hunk 3]
```diff
@@ -47,29 +44,26 @@ contract InterestConsensus is Consensus, Initializable, InterestConsensusInterfa
...
         lenders = lendersAddress;
-        requiredSubmissions = initRequiredSubmissions;
-        maximumTolerance = initMaximumTolerance;
-        responseExpiryLength = initResponseExpiry;
     }
 
     function processRequest(
         ZeroCollateralCommon.InterestRequest calldata request,
         ZeroCollateralCommon.InterestResponse[] calldata responses
     ) external isInitialized() isLenders() returns (uint256) {
```
Comment (hensha256) [inline contracts/base/InterestConsensus.sol:62]: Shuold this be whenNotPaused()?

File: contracts/base/Lenders.sol [hunk 4]
```diff
@@ -64,28 +64,29 @@ contract Lenders is LendersInterface {
...
         interestConsensus = InterestConsensusInterface(interestConsensusAddress);
     }
 
-    /** External Functions */
-
     function setAccruedInterest(
         ZeroCollateralCommon.InterestRequest calldata request,
         ZeroCollateralCommon.InterestResponse[] calldata responses
-    ) external {
+    ) external isInitialized() whenNotPaused() whenLendingPoolNotPaused() nonReentrant() {
```
Comment (hensha256) [inline contracts/base/Lenders.sol:89]: For what reason should this function be `whenLendingPoolNotPaused`? This function doesnt use the lending pool

File: contracts/base/Lenders.sol [hunk 5]
```diff
@@ -118,6 +119,10 @@ contract Lenders is LendersInterface {
         external
         isLendingPool()
         isValid(recipient)
+        isInitialized()
+        whenNotPaused()
+        whenLendingPoolNotPaused()
```
Comment (hensha256) [inline contracts/base/Lenders.sol:124]: As above

Comment (hensha256) [inline contracts/base/Lenders.sol:124]: This function is marked as `isLendingPool`. So if the lending pool is paused, it cannot be called from anywhere. Therefore it doesnt need the `whenLendingPoolNotPaused`, as the lending pool will deal with the pausing

File: contracts/base/Settings.sol [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+// Commons
+import "../util/AddressLib.sol";
+
+// Interfaces
+
+import "../interfaces/SettingsInterface.sol";
+
+
+contract Settings is Pausable, Ownable, SettingsInterface {
```
Comment (hensha256) [inline contracts/base/Settings.sol:30]: Do we need both `Pausable` and `Ownable`? `Pausable` already includes a role

File: test/base/BaseInitializeTest.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+// Smart contracts
+const Base = artifacts.require("./mock/base/BaseMock.sol");
+
+contract('BaseInitializeTest', function (accounts) {
+
+    withData({
+        _1_basic: [accounts[0], undefined, false],
+        _2_notSettings: [NULL_ADDRESS, 'SETTINGS_MUST_BE_PROVIDED', true],
+    }, function(settingsAddress, expectedErrorMessage, mustFail) {
+        it(t('user', 'initialize', 'Should (or not) be able to inintialize the new instance.', mustFail), async function() {
```
Comment (hensha256) [inline test/base/BaseInitializeTest.js:16]: Typo `inintialize`

File: test/base/InterestConsensusInitializeTest.js [hunk 8]
```diff
@@ -1,37 +1,36 @@
...
 contract('InterestConsensusInitializeTest', function (accounts) {
 
     withData({
-        _1_no_reqsubmissions: [true, 0, 1, 'MUST_PROVIDE_REQUIRED_SUBS', true],
-        _2_no_lenders: [false, 1, 1, 'MUST_PROVIDE_LENDER_INFO', true],
-        _3_no_maxtolerance: [true, 1, 0, undefined, false],
-        _4_all_provided: [true, 1, 1, undefined, false],
+        _1_no_settings: [true, false, 'MUST_PROVIDE_SETTINGS', true],
+        _2_no_lenders: [false, true, 'MUST_PROVIDE_LENDER_INFO', true],
+        _4_all_provided: [true, true, undefined, false],
```
Comment (hensha256) [inline test/base/InterestConsensusInitializeTest.js:15]: this should be numbered `3` now

File: contracts/base/Lenders.sol [hunk 9]
```diff
@@ -118,6 +119,10 @@ contract Lenders is LendersInterface {
...
+        nonReentrant()
```
Comment (hensha256) [inline contracts/base/Lenders.sol:125]: why is nonReentrant needed? It doesnt make any external function calls

File: contracts/base/Lenders.sol (hunk 4 above)
Comment (hensha256) [reply contracts/base/Lenders.sol:89]: why is nonReentrant needed? It doesnt make any external function calls

File: contracts/base/InterestConsensus.sol (hunk 3 above)
Comment (salazarguille) [reply contracts/base/InterestConsensus.sol:62]: This function is only called by Lenders.setAccruedInterest. That one contains the modifiers.

File: contracts/base/Lenders.sol (hunk 4 above)
Comment (salazarguille) [reply contracts/base/Lenders.sol:89]: I think if the current lending pool is paused, we shouldn't allow executions on this function.

File: contracts/base/Initializable.sol (hunk 2 above)
Comment (salazarguille) [reply contracts/base/Initializable.sol:46]: it gets conflict with modifier isInitialized (that's why the variable is bool private _isInitialized).

File: contracts/base/Lenders.sol (hunk 5 above)
Comment (hensha256) [reply contracts/base/Lenders.sol]: I don't believe youve made this change?

File: contracts/base/Lenders.sol (hunk 9 above)
Comment (hensha256) [reply contracts/base/Lenders.sol]: Nor this one

File: contracts/base/Lenders.sol (hunk 5 above)
Comment (salazarguille) [reply contracts/base/Lenders.sol]: It's weird. The change was pushed. We will review it after merging.

File: contracts/base/Lenders.sol (hunk 9 above)
Comment (salazarguille) [reply contracts/base/Lenders.sol]: The same. The change was pushed. We will review it after merging.

**Pull request merged**